common/src/main/java/io/pravega/common/io/FileModificationWatcher.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @VisibleForTesting
+    Path directoryOfFileToWatch() {
+        assert this.pathOfFileToWatch != null;
+        return this.pathOfFileToWatch.getParent();
+    }
+
+    @Override
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service for watching file at path: {}", this.pathOfFileToWatch);
+
+            String fileName = nameOfFileToWatch();
+            Path directoryPath = directoryOfFileToWatch();
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.pathOfFileToWatch);
+
+            while (true) {
+                watchKey = watchService.take();
+                log.info("Retrieved and removed watch key for watching file at path: {}", this.pathOfFileToWatch);
+
+                // Looks odd, right? Using the logic or de-duplicating file change events, as suggested by some here:
+                // https://stackoverflow.com/questions/16777869/java-7-watchservice-ignoring-multiple-occurrences-of-
+                // the-same-event
+                Thread.sleep(200);
+
+                if (watchKey != null) {
+                    Optional<WatchEvent<?>> modificationDetectionEvent = watchKey.pollEvents()
+                            .stream()
+                            .filter( // we only care about changes to the specified file.
+                                    event -> event.context().toString().contains(fileName))
+                            .findAny(); // we only care to know whether the
+
+                    if (modificationDetectionEvent.isPresent()) {
+                        log.debug("Detected that the file [{}] has modified", this.pathOfFileToWatch);
+                        callback.accept(modificationDetectionEvent.get());
+                    }
+
+                    boolean isKeyValid = watchKey.reset();
+                    log.debug("Done resetting watch key, so that it can receive further event notifications.");
+                    if (!isKeyValid) {
+                        log.info("No longer watching file [{}]", this.pathOfFileToWatch);
+                        break;
+                    }
+                } else {
+                    log.debug("watchKey for file at path {} was null", this.pathOfFileToWatch);
+                }
+                if (!loopContinuously) {
+                    break;
+                }
+            }
+        } catch (InterruptedException e) {
+            log.info("Thread {}, watching for modifications in file {}, interrupted with cause {}",
```

Comment:
This is not an appropriate way to handle an interrupted exception.
See `Exceptions.handleInterupted`

Reply:
I've modified this part of the code a bit. Please take a look again and see if it looks OK.